src/grammar.rs

```diff
@@ -685,6 +690,158 @@ mod tests {
+        );
+        assert!(
+            right_recursive
+                .parse::<Grammar>()
+                .unwrap()
+                .parse_input(input)
+                .next()
+                .is_some(),
+            "right-recursive grammar failed to parse: {input}"
+        );
+    }
+
+    #[test]
+    fn shared_nullable_nonterminal() {
+        let fails: &str = "
+        <disjunction> ::= <predicate> | <disjunction> <or> <predicate>
+        <predicate> ::= <char_null_one> | <special-string> '.'
+
+        <char_null_one> ::= <char_null_two>
+        <char_null_two> ::= <char_null_three>
+        <char_null_three> ::= <char>
+
+        <or> ::= <ws> 'or' <ws>
+        <ws> ::= <whitespace> | ' ' <ws>
+        <whitespace> ::= ' '
+
+        <special-string> ::= <special-char> | <special-char> <special-string>
+        <special-char> ::= <char> | <whitespace>
+        <char> ::= 'a'
+        ";
+
+        let input = "a or a";
+
+        let passes_1 = fails.replace(
+            // skip nullable production <char_null_two>
+            "<char_null_one> ::= <char_null_two>",
+            "<char_null_one> ::= <char_null_three>",
+        );
+        assert!(passes_1
+            .parse::<Grammar>()
+            .unwrap()
+            .parse_input(input)
+            .next()
+            .is_some());
+
+        let passes_2 = fails.replace(
+            // replace <whitespace> with its terminal ' '
+            "<ws> ::= <whitespace> | ' ' <ws>",
+            "<ws> ::= ' ' | ' ' <ws>",
+        );
+        assert!(passes_2
+            .parse::<Grammar>()
+            .unwrap()
+            .parse_input(input)
+            .next()
+            .is_some());
+
+        let passes_3 = fails.replace(
+            // again, replace <whitespace> with its terminal ' '
+            "<special-char> ::= <char> | <whitespace>",
+            "<special-char> ::= <char> | ' '",
+        );
+        assert!(passes_3
+            .parse::<Grammar>()
+            .unwrap()
+            .parse_input(input)
+            .next()
+            .is_some());
+
+        assert!(
+            fails
+                .parse::<Grammar>()
+                .unwrap()
+                .parse_input(input)
+                .next()
+                .is_some(),
+            "all grammars except last one parsed: {input}"
+        );
+    }
+
+    #[test]
+    fn branching_and_overlapping_parse_trees() {
+        let bnf = "
+        <and> ::= <and> ' AND ' <terminal>
+                | <and> ' ' <terminal>
+                | <terminal>
+        <terminal> ::= 'AND'
+        ";
+        let input = "AND AND AND AND AND";
+
+        // 1. 'AND' <and> 'AND' <and> 'AND'
+        // 2. 'AND' <and> 'AND' 'AND' 'AND'
+        // 3. 'AND' 'AND' <and> 'AND' 'AND'
+        // 4. 'AND' 'AND' 'AND' <and> 'AND'
+        // 5. 'AND' 'AND' 'AND' 'AND' 'AND'
+        assert_eq!(
+            bnf.parse::<Grammar>().unwrap().parse_input(input).count(),
+            5
+        );
+    }
```

Comment:
@amascolo are these parse trees what you expect?

```txt
<and> ::= <and> " AND " <terminal>
├── <and> ::= <and> " AND " <terminal>
│   ├── <and> ::= <terminal>
│   │   └── <terminal> ::= "AND"
│   │       └── "AND"
│   ├── " AND "
│   └── <terminal> ::= "AND"
│       └── "AND"
├── " AND "
└── <terminal> ::= "AND"
    └── "AND"

<and> ::= <and> " AND " <terminal>
├── <and> ::= <and> " " <terminal>
│   ├── <and> ::= <and> " " <terminal>
│   │   ├── <and> ::= <terminal>
│   │   │   └── <terminal> ::= "AND"
│   │   │       └── "AND"
│   │   ├── " "
│   │   └── <terminal> ::= "AND"
│   │       └── "AND"
│   ├── " "
│   └── <terminal> ::= "AND"
│       └── "AND"
├── " AND "
└── <terminal> ::= "AND"
    └── "AND"

<and> ::= <and> " " <terminal>
├── <and> ::= <and> " AND " <terminal>
│   ├── <and> ::= <and> " " <terminal>
│   │   ├── <and> ::= <terminal>
│   │   │   └── <terminal> ::= "AND"
│   │   │       └── "AND"
│   │   ├── " "
│   │   └── <terminal> ::= "AND"
│   │       └── "AND"
│   ├── " AND "
│   └── <terminal> ::= "AND"
│       └── "AND"
├── " "
└── <terminal> ::= "AND"
    └── "AND"

<and> ::= <and> " " <terminal>
├── <and> ::= <and> " " <terminal>
│   ├── <and> ::= <and> " AND " <terminal>
│   │   ├── <and> ::= <terminal>
│   │   │   └── <terminal> ::= "AND"
│   │   │       └── "AND"
│   │   ├── " AND "
│   │   └── <terminal> ::= "AND"
│   │       └── "AND"
│   ├── " "
│   └── <terminal> ::= "AND"
│       └── "AND"
├── " "
└── <terminal> ::= "AND"
    └── "AND"

<and> ::= <and> " " <terminal>
├── <and> ::= <and> " " <terminal>
│   ├── <and> ::= <and> " " <terminal>
│   │   ├── <and> ::= <and> " " <terminal>
│   │   │   ├── <and> ::= <terminal>
│   │   │   │   └── <terminal> ::= "AND"
│   │   │   │       └── "AND"
│   │   │   ├── " "
│   │   │   └── <terminal> ::= "AND"
│   │   │       └── "AND"
│   │   ├── " "
│   │   └── <terminal> ::= "AND"
│   │       └── "AND"
│   ├── " "
│   └── <terminal> ::= "AND"
│       └── "AND"
├── " "
└── <terminal> ::= "AND"
    └── "AND"
```

Reply:
Yep, these look like the parse trees I was expecting!

Thanks even for writing them out in the same order as they appeared in https://github.com/shnewto/bnf/issues/117#issuecomment-1379165220